src/page-view.ts

```diff
@@ -0,0 +1,210 @@
+    $prev_pageview_max_content?: number
+    $prev_pageview_max_content_percentage?: number
+}
+
+interface PageViewEventProperties extends ScrollProperties {
+    $pageview_id: string
+    $prev_pageview_pageview_id: string | undefined
+    $prev_pageview_pathname: string | undefined
+}
+
+interface NonPageEventProperties {
+    $pageview_id: string
+}
+export class PageViewManager {
+    _pageViewData: PageViewData | undefined
+    _hasSeenPageView = false
+    _window: Window
+
+    constructor(window: Window) {
+        this._window = window
+    }
+
+    _createPageViewData(): PageViewData {
+        const data = {
+            pageViewId: uuidv7(),
+            pathname: window?.location.pathname,
+        }
+        return data
+    }
+
+    doPageView(): PageViewEventProperties {
+        let prevPageViewData: PageViewData | undefined
+        // if there were events created before the first PageView, we would have created a
+        // pageViewData for them. If this happened, we don't want to create a new pageViewData
+        if (!this._hasSeenPageView) {
+            this._hasSeenPageView = true
+            prevPageViewData = undefined
+            if (!this._pageViewData) {
+                this._pageViewData = this._createPageViewData()
+            }
+        } else {
+            prevPageViewData = this._pageViewData
+            this._pageViewData = this._createPageViewData()
+        }
+
+        // update the scroll properties for the new page, but wait until the next tick
+        // of the event loop
+        setTimeout(this._updateScrollData, 0)
+
+        return {
+            $pageview_id: this._pageViewData.pageViewId,
+            $prev_pageview_pageview_id: prevPageViewData?.pageViewId,
+            $prev_pageview_pathname: prevPageViewData?.pathname,
+            ...this._calculatePrevPageScrollProperties(prevPageViewData),
+        }
+    }
+
+    doPageLeave(): PageViewEventProperties {
+        const prevPageViewData = this._pageViewData
+        if (!this._pageViewData) {
+            this._pageViewData = this._createPageViewData()
+        }
+        const pageViewData = this._pageViewData
+
+        // prevPageViewData and pageViewData should be the same here, it's unlikely that
+        // this._pageViewData was undefined, but in case something weird happened, don't
+        // send wrong data, just leave those fields undefined
+
+        return {
+            $pageview_id: pageViewData.pageViewId,
+            $prev_pageview_pageview_id: prevPageViewData?.pageViewId,
+            $prev_pageview_pathname: prevPageViewData?.pathname,
+            ...this._calculatePrevPageScrollProperties(prevPageViewData),
+        }
+    }
+
+    getNonPageEvent(): NonPageEventProperties {
+        if (!this._pageViewData) {
+            this._pageViewData = this._createPageViewData()
+        }
+        return {
+            $pageview_id: this._pageViewData.pageViewId,
+        }
+    }
```

Comment:
So I didn't know the historical context - mostly kept it to not break anything that I didn't know the details of 😅 

It could be useful as part of a join, as in you could join pageview events with (pageview|pageleave) events, on event1.pageview_id = event2.prev_pageview_id